File: backend/correlator/pkg/config/loader.go
// Package config provides YAML configuration loading for the correlator service.
// Reference: FR-015 (declarative YAML configuration), FR-016 (JSON Schema validation)
package config

import (
	"fmt"
	"os"
	"time"

	"gopkg.in/yaml.v3"
)

// Config represents the correlator service configuration
type Config struct {
	Global      GlobalConfig      `yaml:"global"`
	Correlator  CorrelatorConfig  `yaml:"correlator"`
	NATS        NATSConfig        `yaml:"nats"`
	ClickHouse  ClickHouseConfig  `yaml:"clickhouse"`
	PostgreSQL  PostgreSQLConfig  `yaml:"postgresql"`
	Observability ObservabilityConfig `yaml:"observability"`
}

// GlobalConfig contains settings shared across all services
type GlobalConfig struct {
	Environment string `yaml:"environment"`
	LogLevel    string `yaml:"log_level"`
	Version     string `yaml:"version"`
}

// CorrelatorConfig contains correlator-specific settings
type CorrelatorConfig struct {
	// Deduplication window in seconds (default: 300 = 5 minutes per FR-002a)
	DedupWindowSeconds int `yaml:"dedup_window_seconds"`

	// Scoring weights for multi-layer correlation (FR-003)
	ScoringWeights ScoringWeights `yaml:"scoring_weights"`

	// Registry penalty for authorized MCPs (FR-005)
	RegistryPenalty int `yaml:"registry_penalty"`

	// Max concurrent event processing workers
	MaxWorkers int `yaml:"max_workers"`

	// Batch size for ClickHouse inserts
	BatchSize int `yaml:"batch_size"`

	// Batch timeout for ClickHouse inserts
	BatchTimeout string `yaml:"batch_timeout"`
}

// ScoringWeights defines weights for different signal types
type ScoringWeights struct {
	Endpoint int `yaml:"endpoint"` // Highest weight (typical: 11)
	Judge    int `yaml:"judge"`    // Medium weight (typical: 5)
	Network  int `yaml:"network"`  // Supporting weight (typical: 3)
}

// NATSConfig contains NATS JetStream connection settings
type NATSConfig struct {
	URL             string        `yaml:"url"`
	ConsumerGroup   string        `yaml:"consumer_group"`
	MaxReconnects   int           `yaml:"max_reconnects"`
	ReconnectWait   time.Duration `yaml:"reconnect_wait"`
	Timeout         time.Duration `yaml:"timeout"`
	TLSEnabled      bool          `yaml:"tls_enabled"`
	TLSCertFile     string        `yaml:"tls_cert_file"`
	TLSKeyFile      string        `yaml:"tls_key_file"`
	TLSCAFile       string        `yaml:"tls_ca_file"`
}

// ClickHouseConfig contains ClickHouse connection settings
type ClickHouseConfig struct {
	Host            string        `yaml:"host"`
	Port            int           `yaml:"port"`
	Database        string        `yaml:"database"`
	Username        string        `yaml:"username"`
	Password        string        `yaml:"password"`
	MaxOpenConns    int           `yaml:"max_open_conns"`
	MaxIdleConns    int           `yaml:"max_idle_conns"`
	ConnMaxLifetime time.Duration `yaml:"conn_max_lifetime"`
	TLSEnabled      bool          `yaml:"tls_enabled"`
	TLSCertFile     string        `yaml:"tls_cert_file"`
	TLSKeyFile      string        `yaml:"tls_key_file"`
	TLSCAFile       string        `yaml:"tls_ca_file"`
}

// PostgreSQLConfig contains PostgreSQL connection settings
type PostgreSQLConfig struct {
	Host            string        `yaml:"host"`
	Port            int           `yaml:"port"`
	Database        string        `yaml:"database"`
	Username        string        `yaml:"username"`
	Password        string        `yaml:"password"`
	SSLMode         string        `yaml:"ssl_mode"`
	MaxOpenConns    int           `yaml:"max_open_conns"`
	MaxIdleConns    int           `yaml:"max_idle_conns"`
	ConnMaxLifetime time.Duration `yaml:"conn_max_lifetime"`
}

// ObservabilityConfig contains metrics and logging settings
type ObservabilityConfig struct {
	MetricsPort        int  `yaml:"metrics_port"`
	HealthCheckEnabled bool `yaml:"health_check_enabled"`
	HealthCheckPort    int  `yaml:"health_check_port"`
}

// LoadConfig loads configuration from YAML files
func LoadConfig(configDir string) (*Config, error) {
	config := &Config{}

	// Load global configuration
	globalPath := fmt.Sprintf("%s/global.yaml", configDir)
	if err := loadYAMLFile(globalPath, config); err != nil {
		return nil, fmt.Errorf("failed to load global.yaml: %w", err)
	}

	// Load correlator-specific configuration
	correlatorPath := fmt.Sprintf("%s/correlator.yaml", configDir)
	if err := loadYAMLFile(correlatorPath, config); err != nil {
		return nil, fmt.Errorf("failed to load correlator.yaml: %w", err)
	}

	// Validate configuration
	if err := config.Validate(); err != nil {
		return nil, fmt.Errorf("configuration validation failed: %w", err)
	}

	return config, nil
}

// loadYAMLFile loads a YAML file
func loadYAMLFile(path string, config *Config) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("failed to read file %s: %w", path, err)
	}

	if err := yaml.Unmarshal(data, config); err != nil {
		return fmt.Errorf("failed to parse YAML from %s: %w", path, err)
	}

	return nil
}

// Validate checks if the configuration is valid
func (c *Config) Validate() error {
	if c.Global.Environment == "" {
		return fmt.Errorf("global.environment is required")
	}

	if c.Correlator.DedupWindowSeconds <= 0 {
		return fmt.Errorf("correlator.dedup_window_seconds must be positive")
	}

	if c.Correlator.ScoringWeights.Endpoint <= 0 {
		return fmt.Errorf("correlator.scoring_weights.endpoint must be positive")
	}

	if c.NATS.URL == "" {
		return fmt.Errorf("nats.url is required")
	}

	if c.ClickHouse.Host == "" {
		return fmt.Errorf("clickhouse.host is required")
	}

	if c.PostgreSQL.Host == "" {
		return fmt.Errorf("postgresql.host is required")
	}

	return nil
}

// GetBatchTimeoutDuration returns the batch timeout as a time.Duration
func (c *Config) GetBatchTimeoutDuration() (time.Duration, error) {
	return time.ParseDuration(c.Correlator.BatchTimeout)
}

// Helper methods for main.go compatibility
func (c *Config) GetDedupWindow() time.Duration {
	return time.Duration(c.Correlator.DedupWindowSeconds) * time.Second
}

func (c *Config) GetClickHouseDSN() string {
	return fmt.Sprintf("tcp://%s:%d?database=%s&username=%s&password=%s",
		c.ClickHouse.Host,
		c.ClickHouse.Port,
		c.ClickHouse.Database,
		c.ClickHouse.Username,
		c.ClickHouse.Password,
	)
}

func (c *Config) GetNATSUrl() string {
	return c.NATS.URL
}

func (c *Config) GetRegistryAPIURL() string {
	// TODO: Make this configurable
	return "http://registry-api:8000"
}

func (c *Config) GetRegistryAuthToken() string {
	// TODO: Make this configurable
	return os.Getenv("REGISTRY_AUTH_TOKEN")
}

func (c *Config) GetMetricsPort() string {
	return fmt.Sprintf(":%d", c.Observability.MetricsPort)
}

func (c *Config) GetHealthPort() string {
	return fmt.Sprintf(":%d", c.Observability.HealthCheckPort)
}

// Flattened config structure for easier access
type FlatConfig struct {
	DedupWindow            time.Duration
	WeightEndpoint         int
	WeightJudge            int
	WeightNetwork          int
	RegistryPenalty        int
	ThresholdAuthorized    int
	ThresholdSuspect       int
	ThresholdUnauthorized  int
	ClickHouseURL          string
	ClickHouseDSN          string
	NATSUrl                string
	RegistryAPIURL         string
	RegistryAuthToken      string
	MetricsPort            string
	HealthPort             string
}

// Flatten converts nested config to flat structure
func (c *Config) Flatten() *FlatConfig {
	return &FlatConfig{
		DedupWindow:            c.GetDedupWindow(),
		WeightEndpoint:         c.Correlator.ScoringWeights.Endpoint,
		WeightJudge:            c.Correlator.ScoringWeights.Judge,
		WeightNetwork:          c.Correlator.ScoringWeights.Network,
		RegistryPenalty:        c.Correlator.RegistryPenalty,
		ThresholdAuthorized:    4,  // Default thresholds per FR-003
		ThresholdSuspect:       8,
		ThresholdUnauthorized:  9,
		ClickHouseURL:          fmt.Sprintf("http://%s:%d", c.ClickHouse.Host, c.ClickHouse.Port),
		ClickHouseDSN:          c.GetClickHouseDSN(),
		NATSUrl:                c.GetNATSUrl(),
		RegistryAPIURL:         c.GetRegistryAPIURL(),
		RegistryAuthToken:      c.GetRegistryAuthToken(),
		MetricsPort:            c.GetMetricsPort(),
		HealthPort:             c.GetHealthPort(),
	}
}

// Load is an alias for LoadConfig
func Load(configDir string) (*FlatConfig, error) {
	cfg, err := LoadConfig(configDir)
	if err != nil {
		return nil, err
	}
	return cfg.Flatten(), nil
}

File: backend/registry-api/src/config/loader.py
"""Configuration loader for Registry API service.

Reference: FR-015 (declarative YAML configuration)
"""
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import yaml


@dataclass
class GlobalConfig:
    """Global settings shared across services."""
    environment: str
    log_level: str
    version: str


@dataclass
class RegistryAPIConfig:
    """Registry API service-specific configuration."""
    host: str
    port: int
    reload: bool  # Auto-reload on code changes (dev only)
    workers: int  # Uvicorn workers
    cors_origins: list[str]  # Allowed CORS origins
    jwt_secret: str  # JWT signing secret
    jwt_algorithm: str  # JWT algorithm (default: HS256)
    jwt_expiration_minutes: int  # JWT token expiration


@dataclass
class PostgreSQLConfig:
    """PostgreSQL connection settings."""
    host: str
    port: int
    database: str
    username: str
    password: str
    ssl_mode: str
    max_connections: int
    pool_size: int
    pool_recycle: int  # Recycle connections after N seconds


@dataclass
class ClickHouseConfig:
    """ClickHouse connection settings for findings queries."""
    host: str
    port: int
    database: str
    username: str
    password: str
    tls_enabled: bool
    tls_cert_file: Optional[str] = None
    tls_key_file: Optional[str] = None
    tls_ca_file: Optional[str] = None


@dataclass
class NotificationConfig:
    """Notification delivery settings (FR-025a)."""
    smtp_host: str
    smtp_port: int
    smtp_username: str
    smtp_password: str
    smtp_from_email: str
    webhook_timeout_seconds: int  # Timeout for webhook calls


@dataclass
class ObservabilityConfig:
    """Observability settings."""
    metrics_port: int
    health_check_enabled: bool
    health_check_port: int


@dataclass
class Config:
    """Complete Registry API service configuration."""
    global_config: GlobalConfig
    registry_api: RegistryAPIConfig
    postgresql: PostgreSQLConfig
    clickhouse: ClickHouseConfig
    notification: NotificationConfig
    observability: ObservabilityConfig


def load_config(config_dir: str) -> Config:
    """Load configuration from YAML files.

    Loads global.yaml and registry-api.yaml.

    Args:
        config_dir: Directory containing YAML config files

    Returns:
        Config: Loaded and validated configuration

    Raises:
        ValueError: If configuration is invalid
    """
    config_path = Path(config_dir)

    # Load global configuration
    global_data = load_yaml_file(config_path / "global.yaml")

    # Load registry-api configuration
    registry_api_data = load_yaml_file(config_path / "registry-api.yaml")

    # Merge configurations
    merged = {**global_data, **registry_api_data}

    # Build Config object
    config = Config(
        global_config=GlobalConfig(**merged["global"]),
        registry_api=RegistryAPIConfig(**merged["registry_api"]),
        postgresql=PostgreSQLConfig(**merged["postgresql"]),
        clickhouse=ClickHouseConfig(**merged["clickhouse"]),
        notification=NotificationConfig(**merged["notification"]),
        observability=ObservabilityConfig(**merged["observability"])
    )

    # Validate
    validate_config(config)

    return config


def load_yaml_file(file_path: Path) -> dict:
    """Load a YAML file and return as dictionary.

    Args:
        file_path: Path to YAML file

    Returns:
        dict: Parsed YAML content
    """
    with open(file_path, "r") as f:
        return yaml.safe_load(f)


def validate_config(config: Config) -> None:
    """Validate configuration values.

    Args:
        config: Configuration to validate

    Raises:
        ValueError: If configuration is invalid
    """
    # Validate JWT secret is set
    if not config.registry_api.jwt_secret or config.registry_api.jwt_secret == "changeme":
        if config.global_config.environment == "prod":
            raise ValueError("JWT secret must be changed in production")

    # Validate database connections
    if not config.postgresql.host:
        raise ValueError("PostgreSQL host is required")

    if not config.clickhouse.host:
        raise ValueError("ClickHouse host is required")

    # Validate CORS origins in production
    if config.global_config.environment == "prod":
        if "*" in config.registry_api.cors_origins:
            raise ValueError("Wildcard CORS not allowed in production")

    # Validate SMTP settings for notifications
    if not config.notification.smtp_host:
        raise ValueError("SMTP host is required for notifications")
